src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs

```diff
@@ -1307,6 +1307,27 @@ internal static bool IsGenericTaskType(this TypeSymbol type, CSharpCompilation c
             return namedType.IsCustomTaskType(out builderArgument);
         }
 
+        internal static bool IsAsyncMainReturnType(this TypeSymbol type, CSharpCompilation compilation)
+        {
+            var namedType = type as NamedTypeSymbol;
+            if (type.IsNonGenericTaskType(compilation))
+            {
+                return true;
+            }
+            else if ((object)namedType == null || namedType.Arity != 1)
```

Comment:
@CyrusNajmabadi: This is checking the arity of the type.  That is, `Foo<T, E>` has arity 2.  

However, I've restructured this method to be a bit more clear.